src/RecurringGrantDrop.sol

```diff
@@ -185,26 +199,96 @@ contract RecurringGrantDrop is Ownable2Step{
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof that demonstrates the claimer has a verified World ID
+    /// @dev hashToField function docs are in lib/world-id-contracts/src/libraries/ByteHasher.sol
+    function claimReserved(uint256 timestamp, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof, bytes calldata signature)
+        external
+    {
+        uint256 grantId = grant.calculateId(timestamp);
+        checkClaimReserved(timestamp, receiver, root, nullifierHash, proof, signature);
+
+        nullifierHashes[nullifierHash] = true;
+
+        SafeERC20.safeTransferFrom(token, holder, receiver, grant.getAmount(grantId));
+
+        emit GrantClaimed(grantId, receiver);
+    }
+
+    /// @notice Check whether a claim is valid
+    /// @param timestamp The timestamp of the reservation
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof, array of 8 uint256 elements, demonstrating that the claimer has a verified World ID
+    /// @param signature The off-chain signature of the reservation.
+    function checkClaimReserved(uint256 timestamp, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof, bytes calldata signature)
+        public
+    {
+        uint256 grantId = grant.calculateId(timestamp);
+
+        if (receiver == address(0)) revert InvalidReceiver();
+        if (timestamp > block.timestamp) revert InvalidTimestamp();
+
+        this.checkNullifier(grantId, receiver, root, nullifierHash, proof);
+
+        grant.checkReservationValidity(timestamp);
+
+        address signer = ECDSA.recover(keccak256(abi.encode(timestamp, nullifierHash)), signature);
+        if (!allowedSigners[signer]) revert UnauthorizedSigner();
+
+        worldIdRouter.verifyProof(
+            root,
+            groupId,
+            uint256(keccak256(abi.encodePacked(receiver))) >> 8,
+            nullifierHash,
+            grantId,
+            proof
+        );
+    }
+
+    /// @notice Check whether a nullifier has been used before in the previous contract.
+    function checkNullifier(uint256 grantId, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof) external {
+        if (nullifierHashes[nullifierHash]) revert InvalidNullifier();
+
+        // If this is for a grant before the new deployment, check the previous contract.
+        if (grantId < RELEASE_GRANT_ID) {
```

Comment:
My thinking was that we do this on the app side. So a hacky reservation contract to wrap the old claim contract, and a new deployment (starting at RELEASE_GRANT_ID) that contains none of this.
The wallet app then checks timestamp to figure out whether it should use hacky reservation contract or new claim+reservation contract.

Reply:
Implemented now in `Legacy` version of the contract.